storage/replica_command.go

```diff
@@ -294,27 +312,67 @@ func (r *Replica) EndTransaction(batch engine.Engine, ms *engine.MVCCStats, args
 		// retry error if the commit timestamp isn't equal to the txn
 		// timestamp.
 		if args.Txn.Isolation == proto.SERIALIZABLE && !reply.Txn.Timestamp.Equal(args.Txn.OrigTimestamp) {
-			return reply, proto.NewTransactionRetryError(reply.Txn)
+			return reply, nil, proto.NewTransactionRetryError(reply.Txn)
 		}
 		reply.Txn.Status = proto.COMMITTED
 	} else {
 		reply.Txn.Status = proto.ABORTED
 	}
 
-	// Persist the transaction record with updated status (& possibly timestamp).
-	if err := engine.MVCCPutProto(batch, ms, key, proto.ZeroTimestamp, nil, reply.Txn); err != nil {
-		return reply, err
+	// Resolve any explicit intents. All that are local to this range get
+	// resolved synchronously in the same batch. The remainder is collected
+	// and handed off to asynchronous processing.
+	var externalIntents []proto.Intent
```

Comment:
I'd relocate the externalIntents var down above the for loop for a little more clarity.